File: asherah-ruby/scripts/interop.rb
#!/usr/bin/env ruby

require "base64"

$LOAD_PATH.unshift(File.expand_path("../lib", __dir__))
require "asherah"

def usage!
  warn "Usage: interop.rb <encrypt|decrypt> <partition> <base64>"
  exit 1
end

usage! if ARGV.length < 3

action = ARGV[0]
partition = ARGV[1]
payload = ARGV[2]

config = {
  "ServiceName" => ENV.fetch("SERVICE_NAME", "service"),
  "ProductID" => ENV.fetch("PRODUCT_ID", "product"),
  "Metastore" => ENV.fetch("Metastore", "memory"),
  "KMS" => ENV.fetch("KMS", "static"),
  "EnableSessionCaching" => ENV.fetch("SESSION_CACHE", "1") != "0",
  "Verbose" => false
}

if ENV.key?("CONNECTION_STRING")
  config["ConnectionString"] = ENV["CONNECTION_STRING"]
elsif ENV.key?("SQLITE_PATH")
  config["ConnectionString"] = ENV["SQLITE_PATH"]
end

Asherah.setup(config)
begin
  case action
  when "encrypt"
    data = Base64.strict_decode64(payload)
    json = Asherah.encrypt(partition, data)
    print Base64.strict_encode64(json)
  when "decrypt"
    json = Base64.strict_decode64(payload).force_encoding(Encoding::UTF_8)
    recovered = Asherah.decrypt(partition, json)
    print Base64.strict_encode64(recovered)
  else
    usage!
  end
ensure
  Asherah.shutdown
end

File: interop/php/interop.php
<?php

declare(strict_types=1);

use GoDaddy\Asherah\Asherah;
use GoDaddy\Asherah\AsherahConfig;
use GoDaddy\Asherah\KmsConfig;
use GoDaddy\Asherah\MetastoreConfig;

require_once dirname(__DIR__, 2) . '/asherah-php/vendor/autoload.php';

if ($argc !== 4) {
    fwrite(STDERR, "usage: php interop/php/interop.php <encrypt|decrypt> <partition-id> <base64>\n");
    exit(2);
}

$metastore = getenv('Metastore') ?: 'memory';
$connectionString = getenv('CONNECTION_STRING') ?: getenv('SQLITE_PATH') ?: '';
$kms = getenv('KMS') ?: 'test-debug-static';
$staticMasterKeyHex = getenv('STATIC_MASTER_KEY_HEX') ?: '';

$metastoreConfig = match ($metastore) {
    'sqlite' => MetastoreConfig::sqlite($connectionString),
    'memory' => MetastoreConfig::memory(),
    default => throw new InvalidArgumentException("unsupported PHP interop metastore: {$metastore}"),
};
$kmsConfig = match ($kms) {
    'static' => KmsConfig::static($staticMasterKeyHex),
    'test-debug-static' => $staticMasterKeyHex !== ''
        ? KmsConfig::static($staticMasterKeyHex)
        : KmsConfig::testDebugStatic(),
    default => throw new InvalidArgumentException("unsupported PHP interop KMS: {$kms}"),
};

$config = new AsherahConfig(
    getenv('SERVICE_NAME') ?: 'service',
    getenv('PRODUCT_ID') ?: 'product',
    $metastoreConfig,
    $kmsConfig
);
Asherah::setup($config->withSessionCache(false));

try {
    $payload = base64_decode($argv[3], true);
    if ($payload === false) {
        throw new InvalidArgumentException('payload must be base64');
    }

    if ($argv[1] === 'encrypt') {
        echo base64_encode(Asherah::encryptBytes($argv[2], $payload));
    } elseif ($argv[1] === 'decrypt') {
        echo base64_encode(Asherah::decryptBytes($argv[2], $payload));
    } else {
        fwrite(STDERR, "usage: php interop/php/interop.php <encrypt|decrypt> <partition-id> <base64>\n");
        exit(2);
    }
} finally {
    Asherah::shutdown();
}
